polynote-spark/src/main/scala/polynote/kernel/remote/protocol.scala

```diff
@@ -0,0 +1,156 @@
+package polynote.kernel.remote
+import polynote.kernel._
+import scodec.{Attempt, Codec, Err}
+import scodec.codecs.{Discriminated, Discriminator, byte}
+import scodec.codecs.implicits._
+import polynote.messages._
+import polynote.runtime.{StreamingDataRepr, TableOp}
+import shapeless.{LabelledGeneric, cachedImplicit}
+
+/*
+ * The protocol for communication with remote kernels. Essentially just an ADT version of KernelAPI, with some limitations
+ */
+sealed trait RemoteRequest {
+  def reqId: Int
+}
+
+abstract class RemoteRequestCompanion[T](msgTypeId: Byte) {
+  implicit val discriminator: Discriminator[RemoteRequest, T, Byte] = Discriminator(msgTypeId)
+}
+
+final case class InitialNotebook(reqId: Int, notebook: Notebook) extends RemoteRequest
+object InitialNotebook extends RemoteRequestCompanion[InitialNotebook](-1) {
```

Comment:
Why `-1` and not `0`?

Reply:
TBH it doesn't matter... was just differentiating requests that expect a response from requests that don't.